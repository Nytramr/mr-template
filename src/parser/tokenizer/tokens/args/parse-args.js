import {
  BOOL_TYPE,
  NIL_TYPE,
  NUMBER_TYPE,
  UNKNOWN_TYPE,
  VALUE_TYPE,
  VARIABLE_TYPE,
} from '../../../../constants';
import { getMatchObject } from '../../utils';

const SPLIT_PIPELINE_STRING_REGEX =
  /\s*((?<string>""|``|`(?:[^`]+)`?|"(?:\\"|[^"])+"?)|(?<char>'(?:\\'|[^'])+'?)|(?<symbol>[|,])|(?<assign>:?=)|(?<paren>[()])|(?<unknown>[^\s()|:,=]+))/dgm;

const TOKEN_TYPES = [
  [NUMBER_TYPE, /^[-+]?[0-9]/],
  [VARIABLE_TYPE, /^\$/],
  [VALUE_TYPE, /^\./],
  [BOOL_TYPE, /^(true|false)/],
  [NIL_TYPE, /^nil/],
];

function inferType(text) {
  let tuple = TOKEN_TYPES.find(([_, regexp]) => regexp.test(text));
  return tuple ? tuple[0] : UNKNOWN_TYPE;
}

export function parseArgs(args) {
  if (args) {
    const matches = Array.from(
      args.text.matchAll(SPLIT_PIPELINE_STRING_REGEX),
    ).flatMap((match) => {
      let { groups, indices } = JSON.parse(
        JSON.stringify({ groups: match.groups, indices: match.indices.groups }),
      );

      return Object.entries(groups).map(([key, text]) => {
        let type = key;
        if (key === UNKNOWN_TYPE) {
          type = inferType(text);
        }
        return {
          ...getMatchObject(text, ...indices[key], args.start),
          type,
        };
      });
    });

    return matches;
  }

  return [];
}
