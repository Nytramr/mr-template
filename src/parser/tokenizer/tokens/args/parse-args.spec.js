import { parseArgs } from './parse-args';

describe('args', () => {
  describe('parseArgs', () => {
    describe('different types', () => {
      test.each([
        ['"My raw string"'],
        ['`My quoted string`'],
        ["' '"],
        ['1234'],
        ['-1234'],
        ['12.34'],
        ['-12.34'],
        ['false'],
        ['true'],
        ['.'],
        ['.value'],
        ['.nested.value'],
        ['nil'],
        ['eq'],
        ['myFunction'],
        ['unknown'],
      ])('%s', (text) => {
        let tokenObj = {
          text,
          start: 0,
        };

        let result = parseArgs(tokenObj, { myFunction: () => {} });

        expect(JSON.stringify(result)).toMatchSnapshot();
      });
    });

    describe('combinations', () => {
      test.each([
        ['.Name | trim | lower | capitalize'],
        ['"Hello, " | cat .Name'],
        ['index .Items 0 | lower'],
        ['len .Items | printf "Count: %d"'],
        ['.CreatedAt | date "2006-01-02"'],
        ['.Content | html'],
        ['true | false | myCustomFunc'],
        ['(index .Items 0 | upper) | printf "First Item: %s"'],
        ['"hello" | upper'],
        ['"WORLD" | lower'],
        ['"   GoLang   " | trim'],
        ['"Hello, " | cat "World!"'],
        ['"Golang Templates" | substring 0 6'],
        ['"Go! " | repeat 3'],
        ['123.456 | printf "%.2f"'],
        ['"Good" | cat " Morning" | upper'],
        ['slice "apple" "banana" "cherry" | join ", "'],
        ['"2023-05-22" | date "January 2, 2006"'],
      ])('%s', (text) => {
        let tokenObj = {
          text,
          start: 0,
        };

        let result = parseArgs(tokenObj, { myFunction: () => {} });

        expect(JSON.stringify(result)).toMatchSnapshot();
      });
    });

    describe('variables', () => {
      test.each([
        ['$v = 30'],
        ['$v = "name"'],
        ['$v = .value'],
        ['$v := 30'],
        ['$v := "name"'],
        ['$v := .value'],
      ])('%s', (text) => {
        let tokenObj = {
          text,
          start: 0,
        };

        let result = parseArgs(tokenObj, { myFunction: () => {} });

        expect(JSON.stringify(result)).toMatchSnapshot();
      });
    });
    // describe('', () => {});
  });
});
