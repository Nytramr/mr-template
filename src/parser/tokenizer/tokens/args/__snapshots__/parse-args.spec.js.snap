// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`args parseArgs combinations "   GoLang   " | trim 1`] = `"[{"text":"\\"   GoLang   \\"","start":0,"end":14,"type":"string"},{"text":"|","start":15,"end":16,"type":"symbol"},{"text":"trim","start":17,"end":21,"type":"unknown"}]"`;

exports[`args parseArgs combinations "2023-05-22" | date "January 2, 2006" 1`] = `"[{"text":"\\"2023-05-22\\"","start":0,"end":12,"type":"string"},{"text":"|","start":13,"end":14,"type":"symbol"},{"text":"date","start":15,"end":19,"type":"unknown"},{"text":"\\"January 2, 2006\\"","start":20,"end":37,"type":"string"}]"`;

exports[`args parseArgs combinations "Go! " | repeat 3 1`] = `"[{"text":"\\"Go! \\"","start":0,"end":6,"type":"string"},{"text":"|","start":7,"end":8,"type":"symbol"},{"text":"repeat","start":9,"end":15,"type":"unknown"},{"text":"3","start":16,"end":17,"type":"number"}]"`;

exports[`args parseArgs combinations "Golang Templates" | substring 0 6 1`] = `"[{"text":"\\"Golang Templates\\"","start":0,"end":18,"type":"string"},{"text":"|","start":19,"end":20,"type":"symbol"},{"text":"substring","start":21,"end":30,"type":"unknown"},{"text":"0","start":31,"end":32,"type":"number"},{"text":"6","start":33,"end":34,"type":"number"}]"`;

exports[`args parseArgs combinations "Good" | cat " Morning" | upper 1`] = `"[{"text":"\\"Good\\"","start":0,"end":6,"type":"string"},{"text":"|","start":7,"end":8,"type":"symbol"},{"text":"cat","start":9,"end":12,"type":"unknown"},{"text":"\\" Morning\\"","start":13,"end":23,"type":"string"},{"text":"|","start":24,"end":25,"type":"symbol"},{"text":"upper","start":26,"end":31,"type":"unknown"}]"`;

exports[`args parseArgs combinations "Hello, " | cat "World!" 1`] = `"[{"text":"\\"Hello, \\"","start":0,"end":9,"type":"string"},{"text":"|","start":10,"end":11,"type":"symbol"},{"text":"cat","start":12,"end":15,"type":"unknown"},{"text":"\\"World!\\"","start":16,"end":24,"type":"string"}]"`;

exports[`args parseArgs combinations "Hello, " | cat .Name 1`] = `"[{"text":"\\"Hello, \\"","start":0,"end":9,"type":"string"},{"text":"|","start":10,"end":11,"type":"symbol"},{"text":"cat","start":12,"end":15,"type":"unknown"},{"text":".Name","start":16,"end":21,"type":"value"}]"`;

exports[`args parseArgs combinations "WORLD" | lower 1`] = `"[{"text":"\\"WORLD\\"","start":0,"end":7,"type":"string"},{"text":"|","start":8,"end":9,"type":"symbol"},{"text":"lower","start":10,"end":15,"type":"unknown"}]"`;

exports[`args parseArgs combinations "hello" | upper 1`] = `"[{"text":"\\"hello\\"","start":0,"end":7,"type":"string"},{"text":"|","start":8,"end":9,"type":"symbol"},{"text":"upper","start":10,"end":15,"type":"unknown"}]"`;

exports[`args parseArgs combinations (index .Items 0 | upper) | printf "First Item: %s" 1`] = `"[{"text":"(","start":0,"end":1,"type":"paren"},{"text":"index","start":1,"end":6,"type":"unknown"},{"text":".Items","start":7,"end":13,"type":"value"},{"text":"0","start":14,"end":15,"type":"number"},{"text":"|","start":16,"end":17,"type":"symbol"},{"text":"upper","start":18,"end":23,"type":"unknown"},{"text":")","start":23,"end":24,"type":"paren"},{"text":"|","start":25,"end":26,"type":"symbol"},{"text":"printf","start":27,"end":33,"type":"unknown"},{"text":"\\"First Item: %s\\"","start":34,"end":50,"type":"string"}]"`;

exports[`args parseArgs combinations .Content | html 1`] = `"[{"text":".Content","start":0,"end":8,"type":"value"},{"text":"|","start":9,"end":10,"type":"symbol"},{"text":"html","start":11,"end":15,"type":"unknown"}]"`;

exports[`args parseArgs combinations .CreatedAt | date "2006-01-02" 1`] = `"[{"text":".CreatedAt","start":0,"end":10,"type":"value"},{"text":"|","start":11,"end":12,"type":"symbol"},{"text":"date","start":13,"end":17,"type":"unknown"},{"text":"\\"2006-01-02\\"","start":18,"end":30,"type":"string"}]"`;

exports[`args parseArgs combinations .Name | trim | lower | capitalize 1`] = `"[{"text":".Name","start":0,"end":5,"type":"value"},{"text":"|","start":6,"end":7,"type":"symbol"},{"text":"trim","start":8,"end":12,"type":"unknown"},{"text":"|","start":13,"end":14,"type":"symbol"},{"text":"lower","start":15,"end":20,"type":"unknown"},{"text":"|","start":21,"end":22,"type":"symbol"},{"text":"capitalize","start":23,"end":33,"type":"unknown"}]"`;

exports[`args parseArgs combinations 123.456 | printf "%.2f" 1`] = `"[{"text":"123.456","start":0,"end":7,"type":"number"},{"text":"|","start":8,"end":9,"type":"symbol"},{"text":"printf","start":10,"end":16,"type":"unknown"},{"text":"\\"%.2f\\"","start":17,"end":23,"type":"string"}]"`;

exports[`args parseArgs combinations index .Items 0 | lower 1`] = `"[{"text":"index","start":0,"end":5,"type":"unknown"},{"text":".Items","start":6,"end":12,"type":"value"},{"text":"0","start":13,"end":14,"type":"number"},{"text":"|","start":15,"end":16,"type":"symbol"},{"text":"lower","start":17,"end":22,"type":"unknown"}]"`;

exports[`args parseArgs combinations len .Items | printf "Count: %d" 1`] = `"[{"text":"len","start":0,"end":3,"type":"unknown"},{"text":".Items","start":4,"end":10,"type":"value"},{"text":"|","start":11,"end":12,"type":"symbol"},{"text":"printf","start":13,"end":19,"type":"unknown"},{"text":"\\"Count: %d\\"","start":20,"end":31,"type":"string"}]"`;

exports[`args parseArgs combinations slice "apple" "banana" "cherry" | join ", " 1`] = `"[{"text":"slice","start":0,"end":5,"type":"unknown"},{"text":"\\"apple\\"","start":6,"end":13,"type":"string"},{"text":"\\"banana\\"","start":14,"end":22,"type":"string"},{"text":"\\"cherry\\"","start":23,"end":31,"type":"string"},{"text":"|","start":32,"end":33,"type":"symbol"},{"text":"join","start":34,"end":38,"type":"unknown"},{"text":"\\", \\"","start":39,"end":43,"type":"string"}]"`;

exports[`args parseArgs combinations true | false | myCustomFunc 1`] = `"[{"text":"true","start":0,"end":4,"type":"boolean"},{"text":"|","start":5,"end":6,"type":"symbol"},{"text":"false","start":7,"end":12,"type":"boolean"},{"text":"|","start":13,"end":14,"type":"symbol"},{"text":"myCustomFunc","start":15,"end":27,"type":"unknown"}]"`;

exports[`args parseArgs different types "My raw string" 1`] = `"[{"text":"\\"My raw string\\"","start":0,"end":15,"type":"string"}]"`;

exports[`args parseArgs different types ' ' 1`] = `"[{"text":"' '","start":0,"end":3,"type":"char"}]"`;

exports[`args parseArgs different types . 1`] = `"[{"text":".","start":0,"end":1,"type":"value"}]"`;

exports[`args parseArgs different types .nested.value 1`] = `"[{"text":".nested.value","start":0,"end":13,"type":"value"}]"`;

exports[`args parseArgs different types .value 1`] = `"[{"text":".value","start":0,"end":6,"type":"value"}]"`;

exports[`args parseArgs different types \`My quoted string\` 1`] = `"[{"text":"\`My quoted string\`","start":0,"end":18,"type":"string"}]"`;

exports[`args parseArgs different types -12.34 1`] = `"[{"text":"-12.34","start":0,"end":6,"type":"number"}]"`;

exports[`args parseArgs different types -1234 1`] = `"[{"text":"-1234","start":0,"end":5,"type":"number"}]"`;

exports[`args parseArgs different types 12.34 1`] = `"[{"text":"12.34","start":0,"end":5,"type":"number"}]"`;

exports[`args parseArgs different types 1234 1`] = `"[{"text":"1234","start":0,"end":4,"type":"number"}]"`;

exports[`args parseArgs different types eq 1`] = `"[{"text":"eq","start":0,"end":2,"type":"unknown"}]"`;

exports[`args parseArgs different types false 1`] = `"[{"text":"false","start":0,"end":5,"type":"boolean"}]"`;

exports[`args parseArgs different types myFunction 1`] = `"[{"text":"myFunction","start":0,"end":10,"type":"unknown"}]"`;

exports[`args parseArgs different types nil 1`] = `"[{"text":"nil","start":0,"end":3,"type":"nil"}]"`;

exports[`args parseArgs different types true 1`] = `"[{"text":"true","start":0,"end":4,"type":"boolean"}]"`;

exports[`args parseArgs different types unknown 1`] = `"[{"text":"unknown","start":0,"end":7,"type":"unknown"}]"`;

exports[`args parseArgs variables $v := "name" 1`] = `"[{"text":"$v","start":0,"end":2,"type":"variable"},{"text":":=","start":3,"end":5,"type":"assign"},{"text":"\\"name\\"","start":6,"end":12,"type":"string"}]"`;

exports[`args parseArgs variables $v := .value 1`] = `"[{"text":"$v","start":0,"end":2,"type":"variable"},{"text":":=","start":3,"end":5,"type":"assign"},{"text":".value","start":6,"end":12,"type":"value"}]"`;

exports[`args parseArgs variables $v := 30 1`] = `"[{"text":"$v","start":0,"end":2,"type":"variable"},{"text":":=","start":3,"end":5,"type":"assign"},{"text":"30","start":6,"end":8,"type":"number"}]"`;

exports[`args parseArgs variables $v = "name" 1`] = `"[{"text":"$v","start":0,"end":2,"type":"variable"},{"text":"=","start":3,"end":4,"type":"assign"},{"text":"\\"name\\"","start":5,"end":11,"type":"string"}]"`;

exports[`args parseArgs variables $v = .value 1`] = `"[{"text":"$v","start":0,"end":2,"type":"variable"},{"text":"=","start":3,"end":4,"type":"assign"},{"text":".value","start":5,"end":11,"type":"value"}]"`;

exports[`args parseArgs variables $v = 30 1`] = `"[{"text":"$v","start":0,"end":2,"type":"variable"},{"text":"=","start":3,"end":4,"type":"assign"},{"text":"30","start":5,"end":7,"type":"number"}]"`;
